lib/helpers/load-os-module.js

```diff
@@ -0,0 +1,25 @@
+'use strict';
+
+/**
+ * @typedef {function(): T} DeferModule
+ * @template T
+ */
+
+/**
+ * @param {string} name
+ * @param {{[key: string]: DeferModule<T>}} deferredModules
+ * @return {T}
+ * @template T
+ */
+function loadOsModule(name, deferredModules) {
+  const { platform } = process;
+  if (!(platform in deferredModules)) {
+    throw new Error(`No "${name}" module for platform "${platform}"`);
+  }
+
+  return deferredModules[platform]();
+}
```

Comment:
I'm not convinced this amount of abstraction is justified. How about a thin wrapper around `require` that just prefixes with the value of `process.platform`--something we could use like `requireOSModule('interaction')`?

Reply:
Isn't what you are suggesting more abstract than what I did?

-----

I see two issues with your suggestion:
- How do we pass `require`'s binding or execution context?
- We lose type checking on the non-static `require` call.

### On passing `require` execution context

How are you thinking `requireOSModule` would wrap `require`?

```js
const requireOSModule = osModuleRequire(require);
module.exports = requireOSModule('interaction');
```

```js
module.exports = requireOSModule(require, 'interaction');
```

```js
module.exports = require(osModule('interaction'));
```

```js
module.exports = require(osModule(__filename, 'interaction'));
```

### On type checking

A wrapper like you are suggesting isn't a static import statement. So its hard for type analysis to check it. #38 we can check the types of the modules defined this way. From what I know, we can't do that with any of my above suggestions.

Reply:
> * How do we pass `require`'s binding or execution context?

Do we need to do this? Our current needs would be served by a `requireOSModule` that operates relative to its own `require` function, and under the file structure you're proposing, that would scale to use with additional AT Driver modules.

> * We lose type checking on the non-static `require` call.

How is this different from the implementation of `loadOSModule` proposed here? It seems like the type annotations in this patch (e.g. `/** @type {ATDriverModules.Interaction} */`) address the same problem and would work equally well for `requireOSModule`.

Reply:
> > * How do we pass `require`'s binding or execution context?
> 
> Do we need to do this? Our current needs would be served by a `requireOSModule` that operates relative to its own `require` function, and under the file structure you're proposing, that would scale to use with additional AT Driver modules.

I don't think we need this, for it to execute, but the module would need to be `lib/modules`.

> > * We lose type checking on the non-static `require` call.
> 
> How is this different from the implementation of `loadOSModule` proposed here? It seems like the type annotations in this patch (e.g. `/** @type {ATDriverModules.Interaction} */`) address the same problem and would work equally well for `requireOSModule`.

Current `loadOSModule` can internally determine the type's of the modules. The proposed `requireOSModule` won't be able to do that. Typescript will compare the value returned by `loadOSModule` against the type specified in `@type` and error if the value cannot be converted. When it does that for the value from `requireOSModule`, it will convert from `any`, the universal type, and never produce a conversion error.

Stated another way, Typescript can determine the type of `require('./win32/interaction')` as `ATDriverModules.Interaction`, but not the type of ```require(`./${process.platform}/${moduleName}`)```, which will default to `any`.

We can omit the `@type` in `./lib/modules/interaction` if we want with `loadOSModule`.

Though for that to work I realize I need to update the uses of `@type` to the following:

```js
module.exports = /** @type {ATDriverModules.Session} */ (
  loadOsModule('interaction', {
```

Here's an example error with the above type mismatch.

```sh
> @bocoup/windows-sapi-tts-engine-for-automation@0.0.4 test-types
> tsc -p tsconfig.json

lib/modules/interaction.js:7:29 - error TS2352: Conversion of type 'Interaction' to type 'Session' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Property '"session.new"' is missing in type 'Interaction' but required in type 'Session'.

7 module.exports = /** @type {ATDriverModules.Session} */ (
                              ~~~~~~~~~~~~~~~~~~~~~~~

  lib/modules/types.js:55:2
    55  *   "session.new": ATDriverModules.SessionNewSession,
        ~~~~~~~~~~~~~~~~~
    '"session.new"' is declared here.


Found 1 error in lib/modules/interaction.js:7
```

Here's what it looks like in my editor.

![Screenshot 2024-02-15 at 11 39 29 AM](https://github.com/w3c/aria-at-automation-driver/assets/833298/47b3d5e9-d0e9-4041-a6a5-0d5328add9c9)

Reply:
Sounds like the type checking justifies this!